design/metadata_kv/index.md

```diff
@@ -0,0 +1,247 @@
+# Design proposal - lakeFS on KV
+
+lakeFS stores 2 types of metadata:
+
+1. Immutable metadata, namely committed entries - stored as ranges and metaranges in the underlying object store)
+2. Mutable metadata, namely IAM entities, branch pointers and uncommitted entries
+
+This proposal describes an alternative implementation for the **mutable** metadata.
+Currently, this type of metadata is stored on (and relies on the guarantees provided by) a PostgreSQL database.
+
+In this proposal, we detail a relatively narrow database abstraction - one that could be satisfied by a wide variety of database systems.
+We also detail how lakeFS operates on top of this abstraction, instead of relying on PostgreSQL specific behavior
+
+## Goals
+
+1. Make it easy to run lakeFS as a managed service in a multitenant environment
+1. Allow running production lakeFS systems on top of a database that operations teams are capable of managing, as not all ops teams are versed in scaling PostgreSQL.
+1. Increase users trust in lakeFS, in terms of positioning: PostgreSQL is apparently not the first DB that comes to mind in relation to scalability
+1. Make lakeFS easier to experiment with: Allow running a local lakeFS server without any external dependencies (e.g. no more `docker-compose` in the quickstart)
+
+## Non-Goals
+
+1. Improve performance, latency or throughput of the system
+1. Provide new features, capabilities or guarantees that weren't previously possible in the lakeFS API/UI/CLI
+
+## Design
+
+At the heart of the design, is a simple Key/Value interface, used for all mutable metadata management.
+
+### Semantics
+
+Order operations by a "happens before" operation: operation A happened before operation B if A
+finished before B started.  If operation A happened before B then we also say that B happened
+after A.  (As usual, these are not a total ordering!)
+
+#### Consistency guarantees
+
+* If a write succeeds, successive reads and listings will return the contents of either that
+  write or of some other write that did not happen before it.
+
+* A succesful read returns the contents of some write that did not happen after it.
+
+* A listing returns the contents of some writes that did not happen after it.
+
+* A successful commit holds the contents of some listing of the entire keyspace.
+
+* Mutating operations (commits and writes) may succeed or fail.  When they fail, their contents
```

Comment:
I feel this item is mostly redundant since it's true for pretty much all object stores and databases that I can think of. perhaps @arielshaqed can clarify?